src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -784,42 +785,69 @@
 
 (defn-validated delete-pending-path-id-orphans!
   "Delete paths in dropped-pids that are no longer mentioned
-  in other factsets."
+   in other factsets."
   [factset-id dropped-pids]
   (when-let [dropped-pids (seq dropped-pids)]
     (let [in-pids (jdbc/in-clause dropped-pids)]
-      (sql/do-prepared
-       (format
-        "DELETE FROM fact_paths fp
-           WHERE fp.id %s
-             AND NOT EXISTS (SELECT 1 FROM facts f
-                               WHERE f.fact_path_id %s
-                                 AND f.fact_path_id = fp.id
-                                 AND f.factset_id <> ?)"
-        in-pids in-pids)
-       (concat dropped-pids dropped-pids [factset-id])))))
+      (if (sutils/postgres?)
+        (sql/do-prepared
+          "DELETE FROM fact_paths fp
+           WHERE fp.id = ANY (?)
+           AND NOT EXISTS (SELECT 1 FROM facts f
+             WHERE f.fact_path_id = ANY (?)
```

Comment:
Took me a bit, but I think I do.  I believe it has to do with the way a correlated subquery like that works.  Without the additional guard, I think it may not correctly match up the subqueries with the candidate delete rows.  Here's (hopefully) an example you can run via psql: https://gist.github.com/rbrw/4c855c8e9c0d42b90e51

If we decide that's correct, I can add some relevant comments to the code this time.